File: javascript/run-length-encoding/run-length-encoding.js
export const encode = (string) => {
  if (string.length < 2)
    return string

  let encoded = ""
  let prev_char = string[0]
  let count = 1
  
  for (const char of string.slice(1)) {
    if (char != prev_char) {
      encoded += count > 1 ? `${count}${prev_char}` : prev_char
      prev_char = char
      count = 0
    }

    count++
  }

  // One last time to get the last character
  encoded += count > 1 ? `${count}${prev_char}` : prev_char
  return encoded
};

export const decode = (string) => {
  if (string.length < 2) 
    return string

  let decoded = ""

  for (const part of string.match(/\d*[a-zA-Z ]/g)) {
    const count = part.match(/\d+/) == null ? 1 : parseInt(part.match(/\d+/)[0])
    const char = part.match(/[a-zA-Z ]/)[0]

    for (let i = 0; i < count; i++) {
      decoded += char
    }
  }

  return decoded
}

File: python/run-length-encoding/run_length_encoding.py
import re

def decode(string):
    if len(string) < 2:
        return string
    decoded = ""
    
    for single in re.findall(r"\d*[a-zA-Z ]", string):
        char = re.search(r"[a-zA-Z ]", single).group()
        count = re.search(r"\d*", single).group()
        if count == "":
            count = "1"

        for i in range(int(count)):
            decoded += char
        
    return decoded

def encode(string):
    if len(string) < 2:
        return string

    encoded = ""
    count = 1
    prev_char = string[0]
    
    for char in string[1:]:
        if char != prev_char:
            encoded += f"{count}{prev_char}" if count > 1 else prev_char
            prev_char = char
            count = 0
        count += 1
    
    # One last time to get the last character
    encoded += f"{count}{prev_char}" if count > 1 else prev_char    
    return encoded
